### PR TITLE
Add support for Mariner Linux on Azure

### LINKED ISSUE #1071: Support Mariner Linux as capi image base
/kind feature

**Describe the solution you'd like**
We are being asked to move our linux base images to mariner linux. https://github.com/microsoft/CBL-Mariner
This ask is to create the Mariner Linux image to be used as the base image for both the controller and worker nodes in a Kubernetes CAPZ cluster. Ideally, this base image is built as part of the CAPI image building infrastructure and is published for use in Azure natively, similar to existing images. 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: N/A
- Kubernetes version: (use `kubectl version`): 1.24.1+ 
- OS (e.g. from `/etc/os-release`): N/A

### PR DESCRIPTION
**Pull request by mboersma**
What this PR does / why we need it:

Adds VHD and SIG build support for [Mariner](https://en.wikipedia.org/wiki/CBL-Mariner) Linux in Azure. See documentation at https://microsoft.github.io/CBL-Mariner/docs/.

Which issue(s) this PR fixes:

Fixes #1071

**Additional context**

I think we should scope this to just Azure support, then follow up with OVA.

### CODE REVIEW

**Comment by mboersma** [comment]
This builds a VHD successfully, but when I make a managed image from that and use it in Cluster API, `kubeadm init` never happens. Any ideas or debugging help would be very welcome!

**Comment by ader1990** [comment]
> This builds a VHD successfully, but when I make a managed image from that and use it in Cluster API, `kubeadm init` never happens. Any ideas or debugging help would be very welcome!

Hello, most likely, cloud-init did not run or did run incorrectly. I think it is needed to make sure that the image has been maybe generalized (aka cleaned from previous runs of cloud-init)?

Cloud-init logs are usually in the two files -> /var/log/cloud*.

File: images/capi/Makefile [hunk 1]
```diff
@@ -625,6 +627,7 @@ build-azure-vhd-ubuntu-1804: ## Builds Ubuntu 18.04 VHD image for Azure
 build-azure-vhd-ubuntu-2004: ## Builds Ubuntu 20.04 VHD image for Azure
 build-azure-vhd-ubuntu-2204: ## Builds Ubuntu 22.04 VHD image for Azure
 build-azure-vhd-centos-7: ## Builds CentOS 7 VHD image for Azure
+build-azure-vhd-mariner-2: ## Builds Mariners 2 VHD image for Azure
```
[the remark this answers is not in the document]
Comment (mboersma) [reply images/capi/Makefile:630]: I'm planning to add it in this one, once I work out the init issues.

File: images/capi/packer/ova/mariner-2.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  "boot_disable_ipv6": "0",
+  "boot_media_path": "/media/HTTP",
+  "build_name": "mariner-2",
+  "distro_arch": "amd64",
+  "distro_name": "cbl-mariner",
+  "distro_version": "2.0",
+  "guest_os_type": "generic-linux",
+  "iso_checksum": "a32238f58f68b348f136c54620bb3a3c60b446ec21fdb9185ecd8979c0fc6e8b",
+  "iso_checksum_type": "sha256",
+  "iso_url": "https://aka.ms/mariner-2.0-x86_64-iso",
```
Comment (ader1990) [inline images/capi/packer/ova/mariner-2.json:12]: Hello,

I have started with this PR and managed to create a Mariner 2 image using QEMU-KVM, with the ISO provided by MSFT. Besides the manual steps I needed to install from ISO (mostly enter + tab that can be eventually automated using boot_command_prefix), the resulting install needs a few more steps to be executed:

 - install cloud-init and jq packages
 - disable firewall (by default only port 22 is allowed in /etc/systemd/scripts/ip4save).
    A K8s node needs port 6443 and other ports allowed, depeding on the services / workloads installed
 - user builder with password builder created | enabled with no sudo pass: echo "builder ALL=(ALL:ALL) NOPASSWD:ALL" >> /etc/sudoers

  Tested the image with CAPI and CAPT (cluster api provider for Tinkerbell) and worked as expected.

Comment (mboersma) [reply images/capi/packer/ova/mariner-2.json:12]: @ader1990 that's great, thanks so much for testing!

I had only stubbed out the boot_command_prefix but I'm planning to flesh it out once I have the other issues fixed. I'm working on opening port 6443 in iptables and hopefully I can address the other things on your list at the same time.

Comment (ader1990) [reply images/capi/packer/ova/mariner-2.json:12]: We have a WiP for the boot command here, should we make a PR on your branch?
https://github.com/mpanduru/image-builder/commit/7220c82dda6e196956a80d3d19713a3a8e012fe5

**Comment by mboersma** [comment]
After figuring out how to open the kube-apiserver port persistently through iptables, this seems to be working, at least for my testing through `tilt` and a customized template pointing to my managed image. ❗ 

I'll do some more testing and clean it up, then look at the -sig and ISO targets.

**Comment by mboersma** [comment]
One thing I can't figure out is why the mariner.yml node task is run twice:

```
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [include_role : node] *****************************************************
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : perform a yum update] ********************************************
    azure-arm.sig-mariner-2: ok: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : install baseline dependencies on Mariner] ************************
    azure-arm.sig-mariner-2: changed: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : install extra rpms] **********************************************
    azure-arm.sig-mariner-2: ok: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : perform a yum update] ********************************************
    azure-arm.sig-mariner-2: ok: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : install baseline dependencies on Mariner] ************************
    azure-arm.sig-mariner-2: changed: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : install extra rpms] **********************************************
    azure-arm.sig-mariner-2: ok: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [node : Ensure overlay module is present] *********************************
    azure-arm.sig-mariner-2: changed: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [node : Ensure br_netfilter module is present] ****************************
    azure-arm.sig-mariner-2: changed: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [node : Persist required kernel modules] **********************************
    azure-arm.sig-mariner-2: changed: [default]
```

I can't see any circular import-type thing, and I'm not even sure if I introduced this. It's benign because the tasks are designed to be idempotent, but shouldn't be happening. Ideas welcome!

**Review by AverageMarcus — COMMENTED** [review]
/lgtm

Only thing I'm not sure about is the naming - is `Mariner 2` the name Azure is publicly using to refer to the OS or is it using `Azure Linux` ?

**Comment by mboersma** [comment]
>  is Mariner 2 the name Azure is publicly using to refer to the OS or is it using Azure Linux ?

It's been called "Mariner" publicly, but recently I saw a press release that it's being rebranded to "Azure Linux" so I ran with that. But I'm not seeing many references to "Azure Linux" yet in the docs, and everyone still calls it "Mariner" AFAICT.

I'd be happy just to revert to "Mariner" everywhere and remove mentions of "Azure Linux" if you think that's clearer. I wasn't 100% clear on which way to go.

File: images/capi/ansible/roles/kubernetes/tasks/mariner.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      - "kubelet-{{ kubernetes_rpm_version }}"
+      - "kubeadm-{{ kubernetes_rpm_version }}"
+      - "kubectl-{{ kubernetes_rpm_version }}"
+      - "kubernetes-cni-{{kubernetes_cni_rpm_version }}"
+
+- name: Allow Kubernetes API server through iptables
+  iptables:
+    chain: INPUT
+    protocol: tcp
+    destination_port: 6443
```
Comment (mboersma) [inline images/capi/ansible/roles/kubernetes/tasks/mariner.yml:40]: Maybe this should be a user var?

And what if a runtime configuration changes the api server listen port from the default 6443? AFAICT the approach [in Photon](https://github.com/kubernetes-sigs/image-builder/blob/main/images/capi/ansible/roles/setup/tasks/photon.yml#L51-L61) is to allow listening on all ports, but that's not very secure.

Also, should this move out of the `kubernetes` role into the `setup` role? That's where other iptables changes seem to be, but I thought this was specific to kube-apiserver so could stay here.

File: images/capi/ansible/roles/containerd/tasks/main.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@
   when: ansible_os_family == "Debian"
 
 - import_tasks: redhat.yml
-  when: ansible_os_family == "RedHat"
+  when: ansible_os_family in ["Common Base Linux Mariner", "RedHat"]
```
Comment (mboersma) [inline images/capi/ansible/roles/containerd/tasks/main.yml:19]: Ansible doesn't recognize Mariner specifically, so `ansible_distribution` and `ansible_os_family` are both "Common Base Linux Mariner". I could fabricate a fact so we can refer to it just as "Mariner", like Flatcar does, if that would be an improvement.

**Comment by AverageMarcus** [comment]
I'm happy with whichever name is going to be most known in the ecosystem. I was just wondering if this would align more with "Amazon Linux" or not.

**Comment by mboersma** [comment]
>  if this would align more with "Amazon Linux" or not.

I'm thinking it might be more appropriate to go with "Mariner" consistently. I only see "Azure Linux" actually being used today to refer to the AKS service-specific images based on Mariner.

**Comment by AverageMarcus** [comment]
Happy to stick with Mariner then! :) 

Mind updating the port to use a user var and then I'm happy to give this an approve :)

**Comment by mboersma** [comment]
/retitle Add support for Mariner Linux on Azure

**Comment by mboersma** [comment]
/hold

I need to verify `LimitNOFILE` as described in [this comment](https://github.com/kubernetes-sigs/image-builder/issues/1071#issuecomment-1518343092).

**Comment by AverageMarcus** [comment]
/lgtm
/approve

Nice work! 😄

**Pull request merged**